libraries/ms-helper/src/persist.c

```diff
@@ -0,0 +1,158 @@
+          persist_commit(persist);
+          break;
+        }
+      }
+    }
+  }
+
+  soft_timer_start_millis(PERSIST_COMMIT_TIMEOUT_MS, prv_periodic_commit, persist, NULL);
+}
+
+StatusCode persist_init(PersistStorage *persist, void *blob, size_t blob_size) {
+  if (blob_size > FLASH_PAGE_BYTES) {
+    return status_code(STATUS_CODE_OUT_OF_RANGE);
+  } else if (blob_size % FLASH_WRITE_BYTES != 0) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  persist->blob = blob;
+  persist->blob_size = blob_size;
+  persist->prev_flash_addr = PERSIST_INVALID_ADDR;
+
+  // Load stored data
+  PersistHeader header = {
+    .marker = PERSIST_VALID_MARKER,  //
+    .size_bytes = 0                  //
+  };
+  persist->flash_addr = PERSIST_BASE_ADDR;
+
+  // Essentially, we want to exit in three scenarios:
+  // * The page has been erased and there are no valid sections. Use the base address.
+  //   Marker == 0xFFFFFFFF, Size == 0xFFFFFFFF
+  // * This is an invalid section. Increment the address by (header + blob).
+  //   Marker != 0xFFFFFFFF, Size != 0xFFFFFFFF
+  // * This is a valid section. Use this address.
+  //   Marker == 0xFFFFFFFF, Size != 0xFFFFFFFF
+  do {
+    status_ok_or_return(
+        flash_read(persist->flash_addr, sizeof(header), (uint8_t *)&header, sizeof(header)));
+    if (header.marker != PERSIST_VALID_MARKER) {
+      persist->flash_addr += sizeof(header) + header.size_bytes;
+    }
+
+    if (persist->flash_addr >= PERSIST_END_ADDR) {
+      // Somehow had zero valid sections remaining - erase page
+      LOG_DEBUG("Somehow started with a full invalid page - erasing page\n");
+      status_ok_or_return(flash_erase(PERSIST_FLASH_PAGE));
+      persist->flash_addr = PERSIST_BASE_ADDR;
+    }
+  } while (header.marker != PERSIST_VALID_MARKER);
+
+  if (header.size_bytes == PERSIST_INVALID_SIZE) {
+    LOG_DEBUG("No valid sections found! New persist data will live at 0x%" PRIx32 "\n",
+              (uint32_t)persist->flash_addr);
+    persist_commit(persist);
+  } else if (header.size_bytes != persist->blob_size) {
+    LOG_DEBUG("Mismatched blob sizes! Invalidating old section\n");
+    persist->prev_flash_addr = persist->flash_addr;
+    persist->flash_addr += sizeof(header) + header.size_bytes;
+    persist_commit(persist);
+  } else {
+    LOG_DEBUG("Found valid section at 0x%" PRIx32 " (0x%" PRIx32 " bytes), loading data\n",
+              (uint32_t)persist->flash_addr, (uint32_t)header.size_bytes);
+    StatusCode ret = flash_read(persist->flash_addr + sizeof(header), persist->blob_size,
+                                (uint8_t *)persist->blob, persist->blob_size);
+    status_ok_or_return(ret);
+
+    // Increment flash_addr to the next new section
+    persist->prev_flash_addr = persist->flash_addr;
+    persist->flash_addr += sizeof(header) + header.size_bytes;
+  }
+
+  return soft_timer_start_millis(PERSIST_COMMIT_TIMEOUT_MS, prv_periodic_commit, persist, NULL);
```

Comment:
You might want to add an API to disable the timer or not start the timer at all. For example on driver controls calibration we don't need to actively be checking for new data unless we have entered a calibration mode. This would just waste resources. It is a nice wrapper around flash that is helpful for storing config or other data that occurs infrequently.

Reply:
Hmm, good point. I can add a periodic enable/disable function.